File: src/vrp/instance.cpp
// Copyright (c) 2025 Laboratory for Combinatorial Optimization in Real-time Environment.
// All rights reserved.

#include "instance.hpp"

#include <iostream>
#include <optional>

#include "rcspp/rcspp.hpp"

Instance::Instance(int nb_vehicles, int capacity, std::optional<std::string> name)
    : nb_vehicles_(nb_vehicles), capacity_(capacity), name_(std::move(name)) {}

const Customer& Instance::add_customer(int customer_id, double pos_x, double pos_y, int demand,
                                       int ready_time, int due_time, int service_time, bool depot) {
    LOG_TRACE(__FUNCTION__, '\n');
    LOG_DEBUG(customer_id,
              ", ",
              pos_x,
              ", ",
              pos_y,
              ", ",
              demand,
              ", ",
              ready_time,
              ", ",
              due_time,
              ", ",
              service_time,
              ", ",
              depot,
              '\n');

    if (depot) {
        depot_customer_id_ = customer_id;
    } else {
        demand_customers_id_.push_back(customer_id);
    }

    customers_by_id_.emplace(
        customer_id,
        Customer(customer_id, pos_x, pos_y, demand, ready_time, due_time, service_time, depot));

    return customers_by_id_.at(customer_id);
}

const std::map<size_t, Customer>& Instance::get_customers_by_id() const {
    LOG_TRACE(__FUNCTION__, '\n');

    return customers_by_id_;
}

const Customer& Instance::get_customer(size_t id) const {
    return customers_by_id_.at(id);
}

const Customer& Instance::get_depot_customer() const {
    LOG_DEBUG("depot_customer_id_=", depot_customer_id_, '\n');

    return customers_by_id_.at(depot_customer_id_);
}

const std::vector<size_t>& Instance::get_demand_customers_id() const {
    return demand_customers_id_;
}

int Instance::get_nb_vehicles() const {
    return nb_vehicles_;
}

int Instance::get_capacity() const {
    return capacity_;
}

File: src/python/vrp/instance.py
#  Copyright (c) 2025 Laboratory for Combinatorial Optimization in Real-time Environment.
#  All rights reserved.

from typing import Optional


class Customer:
    def __init__(
        self,
        id: int,
        pos_x: float,
        pos_y: float,
        demand: int,
        ready_time: int,
        due_time: int,
        service_time: int,
        depot: bool,
    ):
        self.id = id
        self.pos_x = pos_x
        self.pos_y = pos_y
        self.demand = demand
        self.ready_time = ready_time
        self.due_time = due_time
        self.service_time = service_time
        self.depot = depot


class Instance:
    def __init__(self, nb_vehicles: int, capacity: int, name: Optional[str] = None):
        self.__nb_vehicles = nb_vehicles
        self.__capacity = capacity
        self.__name = name
        self.__depot_customer_id: int = 0
        self.__customers_by_id: dict[int, Customer] = {}
        self.__demand_customers_id: list[int] = []

    def add_customer(
        self,
        id: int,
        pos_x: float,
        pos_y: float,
        demand: int,
        ready_time: int,
        due_time: int,
        service_time: int,
        depot: bool,
    ) -> Customer:
        # print("add_customer")
        # print(id, pos_x, pos_y, demand, ready_time, due_time, service_time, depot)

        if depot:
            self.__depot_customer_id = id
        else:
            self.__demand_customers_id.append(id)

        self.__customers_by_id[id] = Customer(
            id, pos_x, pos_y, demand, ready_time, due_time, service_time, depot
        )
        return self.__customers_by_id[id]

    def get_customers_by_id(self) -> dict[int, Customer]:
        return self.__customers_by_id

    def get_customer(self, id: int) -> Customer:
        return self.__customers_by_id[id]

    def get_depot_customer(self) -> Customer:
        return self.__customers_by_id[self.__depot_customer_id]

    def get_demand_customers_id(self) -> list[int]:
        return self.__demand_customers_id

    def get_nb_vehicles(self) -> int:
        return self.__nb_vehicles

    def get_capacity(self) -> int:
        return self.__capacity
    
    def get_name(self) -> str:
        return self.__name
    
    def write_to_file(self, filepath: str) -> None:
        """Écrit l'instance dans un fichier au format Solomon."""
        with open(filepath, "w") as f:
            # Nom de l'instance
            f.write(f"{self.__name or 'INSTANCE'}\n\n")

            # Section véhicules
            f.write("VEHICLE\n")
            f.write(f"NUMBER     CAPACITY\n")
            f.write(f"{self.__nb_vehicles:>8}{self.__capacity:>11}\n\n")

            # Section clients
            f.write("CUSTOMER\n")
            f.write(
                f"{'CUST NO.':>8}{'XCOORD.':>10}{'YCOORD.':>10}"
                f"{'DEMAND':>10}{'READY TIME':>12}{'DUE DATE':>10}{'SERVICE':>10}{'TIME':>7}\n"
            )
            f.write(" \n")

            # Dépôt en premier, puis les clients
            depot = self.get_depot_customer()
            all_ids = [depot.id] + self.__demand_customers_id

            for cid in all_ids:
                c = self.__customers_by_id[cid]
                f.write(
                    f"{c.id:>5}"
                    f"{int(c.pos_x):>9}"
                    f"{int(c.pos_y):>11}"
                    f"{c.demand:>11}"
                    f"{c.ready_time:>12}"
                    f"{c.due_time:>11}"
                    f"{c.service_time:>11}"
                    f"   \n"
                )
